### PR TITLE
Protocol 13

### PR DESCRIPTION
**Pull request by elucidsoft**
## Types of changes
Protocol 13
- [x] New feature (non-breaking change which adds functionality)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [tbd ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: stellar-dotnet-sdk/responses/TransactionResponse.cs [hunk 1]
```diff
@@ -123,17 +123,20 @@ public TransactionResponse()
         }
 
         public TransactionResponse(string hash, uint ledger, string createdAt, string sourceAccount, string feeAccount, bool successful,
-            string pagingToken, long sourceAccountSequence, long feePaid, int operationCount, string envelopeXdr,
+            string pagingToken, long sourceAccountSequence, long maxFee, long feeCharged, int operationCount, string envelopeXdr,
```
Comment (elucidsoft) [inline stellar-dotnet-sdk/responses/TransactionResponse.cs:126]: Isn't this a breaking change? Looks like an additional parameter was added, in order to maintain back compat we should add the new parameter with a default value OR create another constructor with the original set of params, and apply a default value calling this one...

File: stellar-dotnet-sdk/responses/TransactionResponse.cs [hunk 2]
```diff
@@ -57,7 +57,7 @@ public class TransactionResponse : Response, IPagingToken
         public List<string> Signatures { get; private set; }
 
         [JsonProperty(PropertyName = "fee_bump_transaction")]
-        public FeeBumpTransaction FeeBumpTransaction { get; set; }
+        public FeeBumpTransaction FeeBumpTx { get; set; }
```
Comment (elucidsoft) [inline stellar-dotnet-sdk/responses/TransactionResponse.cs:60]: I don't think this rename should occur as it's a breaking change. Should probably leave it as it was...

Comment (elucidsoft) [inline stellar-dotnet-sdk/responses/TransactionResponse.cs:60]: I think we shoul d keep original name here to be consistent with prior name to prevent breaking change.

File: stellar-dotnet-sdk/responses/TransactionResponse.cs [hunk 3]
```diff
@@ -50,7 +51,16 @@ public class TransactionResponse : Response, IPagingToken
...
+        public string ResultMetaXdr { get; private set; }
+
+        [JsonProperty(PropertyName = "signatures")]
+        public List<string> Signatures { get; private set; }
+
+        [JsonProperty(PropertyName = "fee_bump_transaction")]
+        public FeeBumpTransaction FeeBumpTransaction { get; set; }
+
+        [JsonProperty(PropertyName = "inner_transaction")]
+        public InnerTransaction InnerTx { get; private set; }
```
Comment (elucidsoft) [inline stellar-dotnet-sdk/responses/TransactionResponse.cs:63]: Naming should be InnerTransaction for consistency?

File: stellar-dotnet-sdk/responses/TransactionResponse.cs [hunk 4]
```diff
@@ -115,9 +122,10 @@ public TransactionResponse()
             // Used by deserializer
         }
 
-        public TransactionResponse(string hash, uint ledger, string createdAt, string sourceAccount, bool successful,
+        public TransactionResponse(string hash, uint ledger, string createdAt, string sourceAccount, string feeAccount, bool successful,
```
Comment (elucidsoft) [inline stellar-dotnet-sdk/responses/TransactionResponse.cs:125]: see previous code review comment

File: stellar-dotnet-sdk/Transaction.cs [hunk 5]
```diff
@@ -6,28 +6,25 @@
 
 namespace stellar_dotnet_sdk
 {
-    public class Transaction
+    public class Transaction : TransactionBase
     {
-        private const uint BaseFee = 100;
-
-        private Transaction(KeyPair sourceAccount, uint fee, long sequenceNumber, Operation[] operations, Memo memo, TimeBounds timeBounds)
+        public Transaction(IAccountId sourceAccount, uint fee, long sequenceNumber, Operation[] operations, Memo memo, TimeBounds timeBounds)
```
Comment (tamirms) [inline stellar-dotnet-sdk/Transaction.cs:11]: if you want users of the SDK to only be able to create `Transaction` instances using `TransactionBuilder` I think you can do that by making the constructor internal.

File: stellar-dotnet-sdk/Transaction.cs [hunk 6]
```diff
@@ -37,118 +34,91 @@ private Transaction(KeyPair sourceAccount, uint fee, long sequenceNumber, Operat
...
-
         /// <summary>
-        /// Adds a new signature ed25519PublicKey to this transaction.
+        ///     Returns signature base for the given network.
         /// </summary>
-        /// <param name="signer"> signer <see cref="KeyPair"/> object representing a signer</param>
         /// <param name="network">The network <see cref="Network"/> the transaction will be sent to.</param>
-        public void Sign(KeyPair signer, Network network)
+        /// <returns></returns>
+        public override byte[] SignatureBase(Network network)
```
Comment (tamirms) [inline stellar-dotnet-sdk/Transaction.cs:42]: I'm not sure if this will produce the correct hash for v1 transactions. this is the implementation in the Java SDK PR https://github.com/stellar/java-stellar-sdk/pull/278/files#diff-ac19703cf3fa8051734361585adb036cR55 . Can you add tests on `Hash()` for v1 transactions with and without a muxed account as the source account?

File: stellar-dotnet-sdk/TransactionBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            {
+                throw new Exception($"Invalid fee, it should be at least {innerBaseFeeRate} stroops");
+            }
+
+            if (fee < BaseFee)
+            {
+                throw new Exception($"Invalid fee, it should be at least {BaseFee} stroops");
+            }
+
+            var feeBumpFee = fee * (innerOps + 1);
```
Comment (tamirms) [inline stellar-dotnet-sdk/TransactionBuilder.cs:47]: missing overflow check

File: stellar-dotnet-sdk/TransactionBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        /// <summary>
+        ///     Builds a transaction. It will increment sequence number of the source account.
+        /// </summary>
+        /// <returns></returns>
+        public Transaction Build()
+        {
+            var operations = _operations.ToArray();
+
+            var totalFee = operations.Length * _fee;
+            if (totalFee > UInt32.MaxValue) throw new InvalidOperationException("Transaction fee overflow");
```
Comment (tamirms) [inline stellar-dotnet-sdk/TransactionBuilder.cs:127]: I don't think this will detect an overflow since I assume `totalFee ` is a uint32. I think to properly detect an overflow, `totalFee` needs to be a `ulong` and then you can compare against `UInt32.MaxValue`

**Comment by tamirms** [comment]
`ReadChallengeTransaction()` should not allow transactions with memo id addresses as the transaction source account or in the operation source account.

similarly, `BuildChallengeTransaction()` should reject `clientAccountId` if it's a memo id address

File: stellar-dotnet-sdk/Transaction.cs (hunk 5 above)
Comment (fracek) [reply stellar-dotnet-sdk/Transaction.cs:11]: That would be a breaking change. Let's see what @elucidsoft says about it.

File: stellar-dotnet-sdk/TransactionBuilder.cs (hunk 8 above)
Comment (fracek) [reply stellar-dotnet-sdk/TransactionBuilder.cs:127]: `totalFee` is a long. I will rewrite this to be more clear.

File: stellar-dotnet-sdk/Transaction.cs (hunk 5 above)
Comment (elucidsoft) [reply stellar-dotnet-sdk/Transaction.cs:11]: We are trying to avoid breaking changes for this release. So we will probably not do this at this time.  Can should add a deprecate attribute to warn of future deprecation.

File: stellar-dotnet-sdk/TransactionBuilder.cs (hunk 7 above)
Comment (elucidsoft) [reply stellar-dotnet-sdk/TransactionBuilder.cs:47]: Also add a Unit Test to cover overflow scenario.

**Comment by elucidsoft** [comment]
> `ReadChallengeTransaction()` should not allow transactions with memo id addresses as the transaction source account or in the operation source account.
> 
> similarly, `BuildChallengeTransaction()` should reject `clientAccountId` if it's a memo id address

Also need a unit test to cover this scenario as well.

File: stellar-dotnet-sdk/Transaction.cs (hunk 6 above)
Comment (fracek) [reply stellar-dotnet-sdk/Transaction.cs:42]: You're right. I didn't update it.

**Comment by fracek** [comment]
I updated the PR but it will fail. @tamirms can you please check why the hash of the feebump inner tx is correct, but the feebump tx hash is wrong?

**Comment by elucidsoft** [comment]
> I updated the PR but it will fail. @tamirms can you please check why the hash of the feebump inner tx is correct, but the feebump tx hash is wrong?

Resolved

File: stellar-dotnet-sdk/FeeBumpTransaction.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
+using System;
+using stellar_dotnet_sdk.xdr;
+using Int64 = stellar_dotnet_sdk.xdr.Int64;
+
+namespace stellar_dotnet_sdk
+{
+    public class FeeBumpTransaction : TransactionBase
+    {
+        public FeeBumpTransaction(IAccountId feeSource, Transaction innerTx, long fee)
```
Comment (tamirms) [inline stellar-dotnet-sdk/FeeBumpTransaction.cs:9]: should this constructor be internal to encourage use of `BuildFeeBumpTransaction`?

File: stellar-dotnet-sdk/WebAuthentication.cs [hunk 10]
```diff
@@ -98,6 +123,12 @@ public static class WebAuthentication
             if (transaction.SequenceNumber != 0)
                 throw new InvalidWebAuthenticationException("Challenge transaction sequence number must be 0");
 
+            if (!(transaction is Transaction))
```
Comment (tamirms) [inline stellar-dotnet-sdk/WebAuthentication.cs:126]: it seems like this type check is not necessary because transaction is declared to be of type `Transaction`

File: stellar-dotnet-sdk/FeeBumpTransaction.cs (hunk 9 above)
Comment (fracek) [reply stellar-dotnet-sdk/FeeBumpTransaction.cs:9]: I think you're right.

File: stellar-dotnet-sdk/WebAuthentication.cs (hunk 10 above)
Comment (fracek) [reply stellar-dotnet-sdk/WebAuthentication.cs:126]: Correct. I missed that.

**Comment by fracek** [comment]
Updated with the small fixes
